Automattic/wp-calypso
# Gutenlypso: Preserve block attributes when updating media

Our custom Media Modal for Gutenlypso relies on `MediaStore.on( 'change' )` for updating the media items when they are edited within the Media Modal itself. 

But the current approach is triggering an undesired effect: it resets the block attributes to default values (using the `mediaCalypsoToGutenberg` utility) causing the previous values to don't be preserved. 

#### Changes proposed in this Pull Request

* Move the current `change` listener out of the Media Modal, so the callback is executed only once per post, instead of once per block.
* Create new class containing the whole logic that updates the media.
* If the media changes, update the needed block attributes without modifying the rest of them.

#### Know issues

This PR doesn't update the media used as featured media for the post or for a simple payment block. A follow-up PR will cover this.

#### Testing instructions

1. Open the block editor at `/block-editor` and select a site under a paid plan (free plans don't allow to edit images in the Media Modal).
2. Insert an image block and click on the "Media Library" button.
3. Select any image from the Media Modal (or upload a new one and select it).
4. Adjust the image size and enter a caption.
5. Repeat steps 2-4 with other media blocks and selecting the same image as selected in step 3: Gallery, Cover, Media & Text, Tiled Gallery.
6. Select any of the current blocks and click on the Pencil icon on its toolbar.
7. Click on the Edit button in the Media Modal and apply some changes to the image (e.g. flip it).
8. Close the Media Modal without inserting the modified image.
9. Make sure that all the media blocks show the modified image.
10. Check the size and the caption of the previous image block are preserved in both the editor and the post preview.

Editor

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1233880/50563323-2c994b80-0d1c-11e9-8d43-a01c23102ed8.png) | ![image](https://user-images.githubusercontent.com/1233880/50563328-3753e080-0d1c-11e9-8fd7-22467301bb31.png) |

Preview

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1233880/50563331-43d83900-0d1c-11e9-9e1e-c715963beecf.png) | ![image](https://user-images.githubusercontent.com/1233880/50563333-49358380-0d1c-11e9-8cab-ccf3a72d3a8f.png) |


Fixes #29557
